### PR TITLE
Implement the Centaur model

### PR DESCRIPTION
**Pull request by rebryk**
Centaur is TTS model:
- The encoder consists of convolutional layers.
- The decoder consists of pre-net and modified transformer blocks followed by convolutional layers.

### CODE REVIEW

File: open_seq2seq/losses/centaur_loss.py [hunk 1]
```diff
@@ -0,0 +1,188 @@
+from __future__ import absolute_import, division, print_function
+from __future__ import unicode_literals
+
+import tensorflow as tf
+
+from .loss import Loss
+
+
+class CentaurLoss(Loss):
```
Comment (blisc) [inline open_seq2seq/losses/centaur_loss.py]: Is there a way to merge CentaurLoss and TacotronLoss since they are similar? Maybe a base class that they both inherit from?

File: open_seq2seq/models/text2speech_centaur.py [hunk 2]
```diff
@@ -0,0 +1,360 @@
+# Copyright (c) 2018 NVIDIA Corporation
```
Comment (blisc) [inline open_seq2seq/models/text2speech_centaur.py]: Same comment as Centaur Loss, can we merge with Text2Speech?

File: open_seq2seq/parts/centaur/attention.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
+import tensorflow as tf
+
+from open_seq2seq.parts.centaur import ConvBlock
+from open_seq2seq.parts.transformer import attention_layer
+from open_seq2seq.parts.transformer.common import PrePostProcessingWrapper
+from open_seq2seq.parts.transformer.ffn_layer import FeedFowardNetwork
+
+
+class AttentionBlock:
+  def __init__(self,
```
Comment (blisc) [inline open_seq2seq/parts/centaur/attention.py]: Comments about class parameters

File: open_seq2seq/parts/centaur/conv_block.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
+import tensorflow as tf
```
Comment (blisc) [inline open_seq2seq/parts/centaur/conv_block.py]: Can we move this to parts/cnns/conv_blocks.py? Maybe call it casual_conv?

File: open_seq2seq/parts/transformer/attention_layer.py [hunk 5]
```diff
@@ -152,32 +159,56 @@ def call(self, x, y, bias, cache=None):
...
+          mask_large = tf.cast(mask_large, tf.float32)
+          mask_small = tf.sequence_mask(window_pos, maxlen=max_length)
+          mask_small = tf.cast(mask_small, tf.float32)
+          mask = mask_large - mask_small
+          mask = -1e9 * (1 - mask)
+
+          bias = mask + bias
+
+        # Clipping
+        bias = tf.maximum(bias, -1e9)
```
Comment (blisc) [inline open_seq2seq/parts/transformer/attention_layer.py]: What is this clipping for?

File: open_seq2seq/models/text2speech_centaur.py (hunk 2 above)
Comment (rebryk) [reply open_seq2seq/models/text2speech_centaur.py]: Merged.

File: open_seq2seq/losses/centaur_loss.py (hunk 1 above)
Comment (rebryk) [reply open_seq2seq/losses/centaur_loss.py]: Merged.

File: open_seq2seq/decoders/centaur_decoder.py [hunk 6]
```diff
@@ -0,0 +1,452 @@
...
+      return {
+        "outputs": [
+          outputs["spec"], outputs["post_net_spec"], alignments,
+          tf.sigmoid(outputs["stop_token_logits"]), outputs["lengths"], outputs["mag_spec"]
+        ],
+        "stop_token_logits": outputs["stop_token_logits"]
+      }
+
+  @staticmethod
+  def get_required_params():
```
[the remark this answers is not in the document]
Comment (rebryk) [reply open_seq2seq/decoders/centaur_decoder.py]: Done.

File: open_seq2seq/data/text2speech/text2speech.py [hunk 7]
```diff
@@ -426,19 +453,24 @@ def _parse_audio_transcript_element(self, element):
...
-        n_fft=self._n_fft,
-        mag_power=self.params.get('mag_power', 2),
-        feature_normalize=self.params["feature_normalize"],
-        mean=self.params.get("feature_normalize_mean", 0.),
-        std=self.params.get("feature_normalize_std", 1.),
-        trim=self.params.get("trim", False),
-        data_min=self.params.get("data_min", 1e-5),
-        mel_basis=self._mel_basis
-    )
+    if audio_filename in self._cache:
```
[the remark this answers is not in the document]
Comment (rebryk) [reply open_seq2seq/data/text2speech/text2speech.py]: Done.

File: open_seq2seq/parts/centaur/attention.py (hunk 3 above)
Comment (rebryk) [reply open_seq2seq/parts/centaur/attention.py]: Done.

File: open_seq2seq/parts/centaur/conv_block.py (hunk 4 above)
Comment (rebryk) [reply open_seq2seq/parts/centaur/conv_block.py]: We decided not to do it.

File: open_seq2seq/parts/transformer/attention_layer.py (hunk 5 above)
Comment (rebryk) [reply open_seq2seq/parts/transformer/attention_layer.py]: I sum up a few biases. Don't want them to be a large negatives numbers.

File: open_seq2seq/data/text2speech/text2speech.py [hunk 8]
```diff
@@ -51,6 +51,11 @@ def get_optional_params():
             'mel_type': ['slaney', 'htk'],
             "exp_mag": bool,
             'style_input': [None, 'wav'],
+            'n_samples_train': int,
```
[the remark this answers is not in the document]
Comment (rebryk) [reply open_seq2seq/data/text2speech/text2speech.py]: Done.

File: open_seq2seq/models/text2speech_centaur.py (hunk 2 above)
Comment (rebryk) [reply open_seq2seq/models/text2speech_centaur.py]: Merged.
